### PR TITLE
ES|QL: Fix layout management for duplicate stats + mv_expand

### LINKED ISSUE #102120: ES|QL: NullPointerException on stats + mv_expand planning
The problem seems to be related to both multiple aggs with the same expression and mv_expand. Removing one of the two the query works fine

```
show functions |  stats  ctvxDDE = count(*), language_name = count(*), NzAUsvB = count(*) |  mv_expand NzAUsvB
```

```
{
  "error" : {
    "root_cause" : [
      {
        "type" : "null_pointer_exception",
        "reason" : "Cannot read field \"nameIds\" because \"set\" is null",
...
}
```

```
[2023-11-14T09:35:07,032][WARN ][o.e.x.e.a.EsqlResponseListener] [runTask-0] Request failed with status [INTERNAL_SERVER_ERROR]: java.lang.NullPointerException: Cannot read field "nameIds" because "set" is null
        at org.elasticsearch.xpack.esql.planner.Layout$Builder.append(Layout.java:79)
        at org.elasticsearch.xpack.esql.planner.LocalExecutionPlanner.planMvExpand(LocalExecutionPlanner.java:626)
        at org.elasticsearch.xpack.esql.planner.LocalExecutionPlanner.plan(LocalExecutionPlanner.java:202)
        at org.elasticsearch.xpack.esql.planner.LocalExecutionPlanner.planLimit(LocalExecutionPlanner.java:610)
        at org.elasticsearch.xpack.esql.planner.LocalExecutionPlanner.plan(LocalExecutionPlanner.java:200)
        at org.elasticsearch.xpack.esql.planner.LocalExecutionPlanner.planOutput(LocalExecutionPlanner.java:310)
        at org.elasticsearch.xpack.esql.planner.LocalExecutionPlanner.plan(LocalExecutionPlanner.java:224)
        at org.elasticsearch.xpack.esql.planner.LocalExecutionPlanner.plan(LocalExecutionPlanner.java:168)
        at org.elasticsearch.xpack.esql.plugin.ComputeService.runCompute(ComputeService.java:260)
        at org.elasticsearch.xpack.esql.plugin.ComputeService.execute(ComputeService.java:143)
        at org.elasticsearch.xpack.esql.plugin.TransportEsqlQueryAction.lambda$doExecuteForked$3(TransportEsqlQueryAction.java:108)

```

Maybe related to https://github.com/elastic/elasticsearch/issues/102083, but the error is different

### PR DESCRIPTION
**Pull request by luigidellaquila**
Fixes #102120 

In case of multiple aggs with the same expression, the layout has multiple entries pointing to the same position.
MV_EXPAND planning did not manage this situation properly, this change fixes the problem

### CODE REVIEW

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/mv_expand.csv-spec [hunk 1]
```diff
@@ -252,3 +252,20 @@ emp_no:integer  | first_name:keyword | generate_mv:keyword
 10001           | Georgi             | foo
 10002           | Bezalel            | foo
 ;
+
+
+// see https://github.com/elastic/elasticsearch/issues/102120
+expandAfterDuplicateAggs
```
Comment (dnhatn) [inline x-pack/plugin/esql/qa/testFixtures/src/main/resources/mv_expand.csv-spec:258]: @luigidellaquila You might need to append `#[skip:-8.11.99]` to these two tests since they don't work with 8.11.0 or 8.11.1. Similarly to YAML tests, you can adjust the skipped version after backporting.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/Layout.java [hunk 2]
```diff
@@ -108,16 +108,27 @@ public Layout build() {
...
             for (ChannelSet set : channels) {
                 int channel = numberOfChannels++;
-                for (NameId id : set.nameIds) {
-                    ChannelAndType next = new ChannelAndType(channel, set.type);
-                    ChannelAndType prev = layout.put(id, next);
-                    // Do allow multiple name to point to the same channel - see https://github.com/elastic/elasticsearch/pull/100238
-                    // if (prev != null) {
-                    // throw new IllegalArgumentException("Name [" + id + "] is on two channels [" + prev + "] and [" + next + "]");
-                    // }
+                if (set != null) {
```
Comment (costin) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/Layout.java:111]: Please add some comments on why the set could be null - feels like a bug that caused the initialization to be skipped.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/Layout.java [hunk 3]
```diff
@@ -108,16 +108,27 @@ public Layout build() {
...
+                        // if (prev != null) {
+                        // throw new IllegalArgumentException("Name [" + id + "] is on two channels [" + prev + "] and [" + next + "]");
+                        // }
+                    }
                 }
             }
             return new DefaultLayout(Collections.unmodifiableMap(layout), numberOfChannels);
         }
+
+        public void replace(NameId id, NameId id1) {
```
Comment (costin) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/Layout.java:125]: Why is this needed by mv_expand and not other commands?

**Review by costin — COMMENTED** [review]
The PR / code needs some explanation on what's going on behind the scenes and why the set is null and the replacement is needed.
Maybe mv_expand needs to preserve the original attributes instead?

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/Layout.java (hunk 3 above)
Comment (luigidellaquila) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/Layout.java:125]: > Maybe mv_expand needs to preserve the original attributes instead?

It was like this at the beginning, but that's not correct (see [wrong folding](https://github.com/elastic/elasticsearch/issues/101118)).

Before this PR, the layout was rebuilt from scratch replacing the expanded field, but it didn't work properly in cases like this, where the layout.inverse() contains null entries (because multiple projections point to the same layout position, or at least this is my understanding).

This change just makes the layout planing a bit more straight-forward and less error prone, but it doesn't change the previous logic.

> Why is this needed by mv_expand and not other commands?

mv_expand implementation is pretty peculiar since the beginning: it assumed that the input and output layouts were the same (that is not completely true, se above)

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/Layout.java (hunk 2 above)
Comment (luigidellaquila) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/Layout.java:111]: I thought it was a bug as well, but looking at the implementation I'm not completely sure, it could be on purpose

https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/DefaultLayout.java#L42

https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/DefaultLayout.java#L65

maybe @nik9000 has a clue

Anyway, it happens because multiple projections point to the same agg (so to the same channel) and the builder inverse has some empty positions. I'll add a comment

**Comment by astefan** [comment]
I've looked at this in some details and from different angles and I am not convinced this PR is doing the right thing. Let's look at the result of the LogicalPlanOptimizer for a query `row x = 1 | stats a = count(*), b = count(*), c = count(*) | mv_expand c`:

```
Limit[500[INTEGER]]                                                                        = Limit[500[INTEGER]]
\_MvExpand[c{r}#556,c{r}#562]                                                              = \_MvExpand[c{r}#556,c{r}#562]
  \_Aggregate[[],[COUNT(*[KEYWORD]) AS a, COUNT(*[KEYWORD]) AS b, COUNT(*[KEYWORD]) AS c]] !   \_Project[[a{r}#552, a{r}#552 AS b, a{r}#552 AS c]]
    \_Row[[1[INTEGER] AS x]]                                                               !     \_Limit[500[INTEGER]]
                                                                                           !       \_Aggregate[[],[COUNT([2a][KEYWORD]) AS a]]
                                                                                           !         \_Row[[1[INTEGER] AS x]]
```

Especially `\_MvExpand[c{r}#556,c{r}#562]`... why is there a `c{r}` ReferenceAttribute, if "c" in itself is an alias and is regarded as `a{r}#552 AS c`. Shouldn't `c{r}#556` be resolved as `a{r}#552 AS c`? For example, for query `row x = 1 | stats a = count(*), b = count(*), c = count(*) | rename c AS xxx | limit 1 | sort xxx` which I believe is similar to the one that's using `mv_expand`, the optimizer does this:

```
Limit[10000[INTEGER]]                                                                          ! Project[[a{r}#507, a{r}#507 AS b, a{r}#507 AS xxx]]
\_OrderBy[[Order[xxx{r}#513,ASC,LAST]]]                                                        ! \_TopN[[Order[a{r}#507,ASC,LAST]],1[INTEGER]]
  \_Limit[1[INTEGER]]                                                                          =   \_Limit[1[INTEGER]]
    \_EsqlProject[[a{r}#507, b{r}#509, c{r}#511 AS xxx]]                                       !     \_Aggregate[[],[COUNT([2a][KEYWORD]) AS a]]
      \_Aggregate[[],[COUNT(*[KEYWORD]) AS a, COUNT(*[KEYWORD]) AS b, COUNT(*[KEYWORD]) AS c]] !       \_Row[[1[INTEGER] AS x]]
        \_Row[[1[INTEGER] AS x]]                                                               !
```

Why I'm mentioning this? `TopN` in the example above is the only other command that's using the `inverse` method of the `Layout` in `LocalExecutionPlanner.planTopN` and it doesn't have a problem with it.

--- 

On the other hand the `Layout` is defined as 
> Maintains the mapping from attribute ids to channels (block index). An attribute can only be mapped to exactly one channel but one channel can be mapped to multiple attributes.

The `project` that sits before `mv_expand` (`\_Project[[a{r}#552, a{r}#552 AS b, a{r}#552 AS c]]`) will create a Layout of the following form:
image

which seems to indicate that there are three channels each having the same reference attributes pointing at them. When Layout.build is called when planning the projection, it will put next to each nameId the "last" channel and the resulting layout will look like (only channel 2 is referenced everywhere):
image

When `planMvExpand` is called and goes over the "inverse" `for (int index = 0; index < inverse.size(); index++) {` for channel (I am assuming here that the index in the array is the index of the channel - block index, as the Layout defines it) 0 and 1 it won't find anything (or it finds a `null`).
Getting back to `TopN`... it doesn't have an issue because its source layout that's using comes from `limit` and it's of size 1. `mv_expand` gets the layout from `project`... maybe `mv_expand` needs to be pushed down past `project`....

**Comment by astefan** [comment]
What I'm trying to say is that the Layout seems to do some things that don't seem right, but at the same time the manipulation of the NameId to mirror what https://github.com/elastic/elasticsearch/pull/101385 introduced doesn't seem right either.

**Comment by luigidellaquila** [comment]
Thanks for checking @astefan 

> Especially \_MvExpand[c{r}#556,c{r}#562]... why is there a c{r} ReferenceAttribute, if "c" in itself is an alias and is regarded as a{r}#552 AS c. Shouldn't c{r}#556 be resolved as a{r}#552 AS c?

IMHO this is correct. Referencing to `a{r}#552 AS c` alias rather than `c`  can be considered an optimization (not sure it makes a difference in terms of performance though), but both are acceptable. In terms of clarity, I actually prefer referring to `c` as the source for mv_expand, that is what the query indicates.

> The project that sits before mv_expand (\_Project[[a{r}#552, a{r}#552 AS b, a{r}#552 AS c]]) will create a Layout of the following form:

I really think the problem is here, it clearly violates the contract you mentioned above.

> maybe mv_expand needs to be pushed down past project....

that looks a bit like sweeping it under the carpet. Pushdown should be an optimization, it should not hide structural problems, especially because in some circumstances the pushdown could be impossible, and bugs would kick in

I'll try to understand why the problem with the layout happens in the first place, probably that's the real fix we need.

**Comment by luigidellaquila** [comment]
I managed to restrict the problem to a wrong management of the layout in Project planning.
I'm closing this and opening https://github.com/elastic/elasticsearch/pull/102399 with a proper fix

**Comment by astefan** [comment]
> IMHO this is correct. Referencing to a{r}#552 AS c alias rather than c can be considered an optimization (not sure it makes a difference in terms of performance though), but both are acceptable.

The optimizer and its optimization rules are not always "performance" optimizations; also it's not a matter of "clarity". For this particular case it's a matter of correctness and query planning consistency. `mv_expand` is acting on attribute reference `c{r}` which in turn is an alias to attribute reference `a{r}`. If all "copies" of the same attribute `a` are regarded as an alias to `a` why wouldn't `c{r}` be regarded as an alias to the one and only reference attribute `a`? This is a matter for another time.

> that looks a bit like sweeping it under the carpet

When I looked at the plan I asked myself "does it make sense to have the projection after mv_expand? We do this for most of the commands... we let them get passed the projections. The projections are "transparent" to pushing downs in general because they ultimately express the final result and it makes sense to be at the farther end of the query. The commands (except those that actually act on the projected columns) eval, filter, sort, enrich, grok, dissect get pushed down past projection commands. Imo, `mv_expand` could do the same thing.

**Comment by luigidellaquila** [comment]
>  The commands (except those that actually act on the projected columns) eval, filter, sort, enrich, grok, dissect get pushed down past projection commands. Imo, mv_expand could do the same thing.

Good point, it's definitely something we can investigate
